packages/language/src/language-server/document-symbol-builder.ts

```diff
@@ -0,0 +1,329 @@
+      ) {
+        const name = element.name ?? "<unknown>";
+        const range = getSyntaxNodeRange(element) ?? { start: 0, end: 0 };
+        levelSymbols.push({
+          level: item.level ?? inheritedLevel,
+          symbol: createDocumentSymbol(
+            name,
+            kind,
+            range,
+            children,
+            textDocument,
+          ),
+        });
+      }
+    }
+
+    return levelSymbols;
+  }
+}
+
+class LevelHierarchyBuilder {
+  private levelMap: Map<number, DocumentSymbol>;
+  private rootSymbols: DocumentSymbol[];
+
+  constructor() {
+    this.levelMap = new Map();
+    this.rootSymbols = [];
+  }
+
+  addSymbol(level: number | null, symbol: DocumentSymbol): void {
+    if (!level) {
+      this.rootSymbols.push(symbol);
+      return;
+    }
+
+    let parentLevel = level - 1;
+    while (!this.levelMap.has(parentLevel) && parentLevel > 0) {
+      parentLevel--;
+    }
+
+    if (parentLevel > 0) {
+      const parentSymbol = this.levelMap.get(parentLevel);
+      if (parentSymbol) {
+        parentSymbol.children = [symbol, ...(parentSymbol.children ?? [])];
+        parentSymbol.kind = SymbolKind.Struct;
+        parentSymbol.range.end = getLastPositionOrDefault(
+          parentSymbol.children,
+          symbol.range.end,
+        );
+      }
+    } else {
+      this.rootSymbols.push(symbol);
+    }
+
+    this.levelMap.set(level, symbol);
+    // Remove all larger levels to reset position
+    for (const key of this.levelMap.keys()) {
+      if (key > level) {
+        this.levelMap.delete(key);
+      }
+    }
+  }
+
+  getRootSymbols(): DocumentSymbol[] {
+    return this.rootSymbols;
+  }
+}
+
+class LabelSymbolBuilder implements SymbolBuilder {
+  canHandle(token: IToken): boolean {
+    if (
+      token.payload.kind !== CstNodeKind.LabelPrefix_Name ||
+      token.payload?.element.kind !== SyntaxKind.LabelPrefix
+    ) {
+      return false;
+    }
+
+    const container = token.payload?.element.container;
+    if (!container || container.kind !== SyntaxKind.Statement) {
+      return true;
+    }
+
+    return container.value?.kind !== SyntaxKind.ProcedureStatement;
+  }
+
+  buildSymbols(
+    token: IToken,
+    elementTokens: IToken[],
+    textDocument: TextDocument,
+    childSymbols: DocumentSymbol[],
+  ): DocumentSymbol[] {
+    const labelPrefixStatement = token.payload?.element.container as Statement;
+    if (!labelPrefixStatement?.labels?.length) {
+      return [];
+    }
+
+    const documentSymbols: DocumentSymbol[] = [];
+    for (const label of labelPrefixStatement.labels) {
+      const name = label.name ?? "<unknown>";
+      const range = getSyntaxNodeRange(label) ?? {
+        start: token.startOffset,
+        end: token.endOffset ?? token.startOffset,
+      };
+
+      documentSymbols.push(
+        createDocumentSymbol(
+          name,
+          SymbolKind.Key,
+          range,
+          childSymbols,
+          textDocument,
+        ),
+      );
```

Comment:
Suggestion: Thinking about this again, will `getSyntaxNodeRange` ever return `undefined` for a valid token? And if the token is invalid, shouldn't we prevent the creation of the document symbol? IMO:
1. `getSyntaxNodeRange` should also check for `isValidToken` and return `undefined` if the token is not valid
2. If `getSyntaxNodeRange` returns `undefined`, skip the respective element, since we cannot navigate to it anyway.